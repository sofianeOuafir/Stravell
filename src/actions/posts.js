import database from "../firebase/firebase";

const addPost = post => ({
  type: "ADD_POST",
  post
});

const startAddPost = (postData = {}) => {
  return (dispatch, getState) => {
    const { uid, userName, userPhotoURL } = getState().auth;
    const { title, body, description, image, createdAt, updatedAt } = postData;
    const post = {
      uid,
      userName,
      userPhotoURL,
      title,
      description,
      image,
      body,
      createdAt,
      updatedAt
    };
    return database
      .ref("posts")
      .push(post)
      .then(ref => {
        dispatch(
          addPost({
            id: ref.key,
            ...post
          })
        );
      })
      .catch(e => console.log(e));
  };
};

const editPost = ({ id, updates }) => ({
  type: "EDIT_POST",
  id,
  updates
});

const startEditPost = ({ id, updates }) => {
  return dispatch => {
    database
      .ref(`posts/${id}`)
      .update(updates)
      .then(ref => {
        dispatch(editPost({ id, updates }));
      });
  };
};

const setPosts = posts => ({
  type: "SET_POSTS",
  posts
});

const startSetPosts = () => {
  return dispatch => {
    return new Promise((resolve, reject) => {
      let posts = [];
      database
      .ref("posts")
      .orderByChild("createdAt")
      .on("value", snapshot => {
        snapshot.forEach(snapshotChild => {
          posts.push({
            id: snapshotChild.key,
            ...snapshotChild.val()
          });
        });
        posts = posts.reverse();
        dispatch(setPosts(posts));
        resolve();
      });
      
    })
  };
};

export { startEditPost, startAddPost, startSetPosts };
