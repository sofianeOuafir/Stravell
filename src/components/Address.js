import React from "react";
import { MdMyLocation } from "react-icons/md";
import Link from "next/link";
import { slugify } from "underscore.string";

import { getPlaceIdFromLatLng } from "./../lib/utils/place";

const Address = ({ iconClassName, addressClassName, address, lat, lng }) => {
  const placeId = getPlaceIdFromLatLng({ lat, lng });
  return (
    <div className="flex align-items--center">
      <Link
        href={`/place?id=${placeId}`}
        as={`/place/${slugify(address)}/${placeId}`}
      >
        <div>
          <MdMyLocation className={iconClassName ? iconClassName : ""} />
          <span className={addressClassName ? addressClassName : ""}>
            {address}
          </span>
        </div>
      </Link>
    </div>
  );
};

export default Address;
