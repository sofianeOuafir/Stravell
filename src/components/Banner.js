import React from "react";
import { IoIosSearch, IoIosAirplane } from "react-icons/io";
import Typed from "react-typed";
import { Link as ScrollLink } from "react-scroll";

import SocialShareButtons from "./SocialShareButtons";
import Image from './Image';
class Banner extends React.Component {
  render() {
    const { imageUrl, place, placeSuggestions } = this.props;
    return (
      <div className="banner__image-container">
        <div className="content-container">
          <div className="banner__text-container" />
          <div className="banner__text">
            <h1 className="banner__website-name">
              Stravell {place ? `| ${place}` : ""}
            </h1>
            <span className="banner__description">
              A search engine for travel articles. <br /> It makes it easy for you
              to find travel articles about thousands of places
              <br />
              around the world written by some of the best travel bloggers
              worldwide.
            </span>

          </div>
          <div className="banner__typing-container" />
          <div className="banner__typing flex align-items--center">
            <IoIosSearch className="mr1" />
            <Typed
              strings={placeSuggestions}
              typeSpeed={100}
              backSpeed={100}
              loop
            />
          </div>
          <div className="banner__get_started-container">
            <ScrollLink
              href="#"
              className="button button--large button--warm-peach"
              to="searchBarInput"
              smooth="true"
              duration={600}
              offset={-100}
            >
              <div className="flex align-items--center">
                <IoIosAirplane className="mr1" /> <span>Get Started</span>
              </div>
            </ScrollLink>
          </div>

          <div className="banner__social_buttons">
            <SocialShareButtons />
          </div>
        </div>
        <Image className="fullwidth" src={imageUrl} alt="" loadingImageWidth={1026} loadingImageHeight={405} />
      </div>
    );
  }
}

export default Banner;
