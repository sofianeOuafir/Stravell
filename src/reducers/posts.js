const postsReducer = (state = [], action) => {
  switch (action.type) {
    case "ADD_POST":
      return [action.post, ...state];
    case "EDIT_POST":
      return [
        ...state.map(post => {
          if (post.id === action.id) {
            return {
              ...post,
              ...action.updates
            };
          } else {
            return post;
          }
        })
      ];
    case "SET_POSTS":
      return [
        ...action.posts.map(post => {
          if (post.comments) {
            let comments = [];
            for (let i in post.comments) {
              comments.push(post.comments[i]);
            }
            post.comments = comments;
          }
          return post;
        })
      ];
    case "EDIT_POSTS_USER_PHOTO_URL":
      return [
        ...state.map(post => {
          if (post.uid == action.uid) {
            return {
              ...post,
              userPhotoURL: action.userPhotoURL
            };
          } else {
            return post;
          }
        })
      ];
    case "ADD_POST_COMMENT":
      return [
        ...state.map(post => {
          if (post.id == action.comment.postId) {
            if (!post.comments) {
              post.comments = [];
            }
            // post.comments = post.comments.push(action.comment);
            post.comments = [...post.comments, action.comment];
            return post;
          } else {
            return post;
          }
        })
      ];
    default:
      return state;
  }
};

export default postsReducer;
