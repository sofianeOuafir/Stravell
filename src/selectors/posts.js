const getVisiblePosts = (posts, { text, countryCode }) => {
  text = text.toLowerCase()
  return posts.filter(post => {
    const matchText =
      post.title.toLowerCase().includes(text) ||
      post.description.toLowerCase().includes(text);
    const matchCountry = countryCode === '' || post.countryCode === countryCode
    return matchText && matchCountry;
  });
};

export { getVisiblePosts };
