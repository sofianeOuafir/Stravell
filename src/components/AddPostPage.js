import React from "react";
import PostForm from "./PostForm";
import { connect } from "react-redux";
import { startAddPost } from "../actions/posts";
import PageHeader from './PageHeader';

class AddPostPage extends React.Component {
  onSubmit = post => {
    this.props.startAddPost(post);
    this.props.history.push("/dashboard");
  };

  render() {
    return (
      <div>
        <PageHeader title="Create Post" />
        <div className="content-container">
          <PostForm onSubmit={this.onSubmit} /> 
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddPost: post => {
    dispatch(startAddPost(post));
  }
});

export default connect(
  undefined,
  mapDispatchToProps
)(AddPostPage);
