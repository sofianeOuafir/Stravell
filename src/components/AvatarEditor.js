import React, { Fragment } from "react";
import Editor from "react-avatar-editor";
import { connect } from "react-redux";
import uuid from "uuid";
import Avatar from "react-avatar";
import { IoMdCamera } from "react-icons/io";
import Modal from "react-modal";

import { uploadFile } from "./../aws/s3";
import { updateUserProfilePicture } from "./../queries/user";
import { updateProfilePicture } from "./../actions/auth";
import Loading from "./Loading";

class AvatarEditor extends React.Component {
  constructor(props) {
    super(props);
    const { userPhotoURL: currentImageUrl, uid } = props.user;
    this.state = {
      currentImageUrl,
      uploadedImage: null,
      uid,
      uploading: false,
      modalIsOpen: false
    };
  }

  closeModal = () => {
    this.setState(() => ({ modalIsOpen: false }));
  };

  onImageSave = () => {
    const { uid } = this.state;
    const { updateProfilePicture } = this.props;
    const canvas = this.editor.getImage();
    this.setState(() => ({ uploading: true, modalIsOpen: false }));

    canvas.toBlob(blob => {
      const file = new File([blob], "profile_picture", {
        type: blob.type
      });
      uploadFile({ file, location: `users/${uid}` })
        .then(data => {
          const { Location: newUserPhotoURL } = data;
          this.setState(() => ({
            uploading: false,
            currentImageUrl: `${newUserPhotoURL}?${new Date().getTime()}`
          }));

          updateUserProfilePicture({ uid, newUserPhotoURL });
          updateProfilePicture(`${newUserPhotoURL}?${new Date().getTime()}`);
        })
        .catch(e => {
          this.setState(() => ({ uploading: false }));
          alert(e);
        });
    });
  };

  onImageChange = e => {
    const uploadedImage = e.target.files[0];
    if (!uploadedImage) {
      return;
    }
    this.setState(() => ({ uploadedImage, modalIsOpen: true }));
  };

  onCancel = () => {
    this.setState(() => ({
      modalIsOpen: false
    }));
  };

  triggerFileDialog = () => {
    document.getElementById("imageInput").click();
  };

  setEditorRef = editor => (this.editor = editor);

  render() {
    const {
      uploadedImage,
      currentImageUrl,
      uploading,
      modalIsOpen
    } = this.state;
    return (
      <div
        className="relative flex align-items--center justify-content--center"
        style={{ height: 100, width: 100 }}
      >
        <Fragment>
          <Avatar
            className="absolute pointer"
            src={currentImageUrl}
            round={true}
            size={100}
            onClick={this.triggerFileDialog}
          />
          {uploading ? (
            <div className="absolute">
              <Loading size={40} />
            </div>
          ) : (
            <Fragment>
              <input
                id="imageInput"
                className="hide"
                type="file"
                onChange={this.onImageChange}
              />

              <IoMdCamera
                className="absolute pointer h2 text-white"
                onClick={this.triggerFileDialog}
              />
            </Fragment>
          )}
        </Fragment>

        <Modal
          isOpen={modalIsOpen}
          onRequestClose={this.closeModal}
          className="modal"
          contentLabel="Example Modal"
        >
          <p className="modal__title h3 favourite-font-weight">Edit Media</p>
          <Editor ref={this.setEditorRef} image={uploadedImage} />
          <div className="flex justify-content--between">
            <button className="button" onClick={this.onImageSave}>Save</button>
            <button className="button button--warm-peach" onClick={this.onCancel}>Cancel</button>
          </div>
        </Modal>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  updateProfilePicture: userPhotoURL =>
    dispatch(updateProfilePicture(userPhotoURL))
});

export default connect(
  null,
  mapDispatchToProps
)(AvatarEditor);
