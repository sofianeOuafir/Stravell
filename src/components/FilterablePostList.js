import React from "react";
import { connect } from "react-redux";
import SearchBar from "./SearchBar";
import PostList from "./PostList";
import { getVisiblePosts } from "./../selectors/posts";

export class FilterablePostList extends React.Component {
  constructor(props) {
    super(props);
  }

  getNoPostText() {
    if(this.props.posts.length === 0) {
      return this.props.noPostText;
    } else if (this.props.filteredPosts.length === 0) {
      return `No results were found for ${this.props.filters.text}.`;
    }
  }

  render () {
    return (
      <div>
        {this.props.posts.length > 0 && <SearchBar autoFocus={this.props.SearchBarAutoFocus} />}
        <PostList
          editable={this.props.editable}
          className="post-list post-list--no-border-top"
          posts={this.props.filteredPosts}
          noPostText={this.getNoPostText()}
        />
      </div>
    );
  }
}


const mapStateToProps = ({ filters }, { posts }) => {
  return {
    posts: posts,
    filteredPosts: getVisiblePosts(
      posts,
      filters
    ),
    filters
  }
};

export default connect(
  mapStateToProps
)(FilterablePostList);
