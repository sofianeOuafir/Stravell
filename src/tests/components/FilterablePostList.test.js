import React from 'react';
import { shallow } from 'enzyme';
import { FilterablePostList } from './../../components/FilterablePostList';
import posts from './../fixtures/posts';

describe('the component get passed a non empty array of posts', () => {
  test('should render correctly', () => {
    const wrapper = shallow(<FilterablePostList posts={posts} filteredPosts={posts} />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('the component get passed an empty array of posts', () => {
  test('should render correctly and not display the filters component', () => {
    const posts = [];
    const wrapper = shallow(<FilterablePostList posts={posts} filteredPosts={posts} />);
    expect(wrapper).toMatchSnapshot();
  });
});
