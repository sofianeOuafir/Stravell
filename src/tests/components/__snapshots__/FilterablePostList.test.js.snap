// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the component get passed a non empty array of posts should render correctly 1`] = `
<div>
  <Connect(SearchBar) />
  <PostList
    className="post-list post-list--no-border-top"
    posts={
      Array [
        Object {
          "body": "{\\"blocks\\":[{\\"key\\":\\"f6q0s\\",\\"text\\":\\"It's amazing how many people, when they think of travel, automatically assume that you must travel across lands in order to expierence a rich beauty or space that you have not yet seen. I am originally from Ontario, Canada and when meeting other travellers from around the world, they would always express how lucky I was to be from such a beautiful country. I always could agree, as I knew that many places all over Canada, some even I haven't been have beautfiul landscapes. What I always seem to forget, is how beautiful my home province of Ontario, truly is. Recently, a friend from another country had come to visit Canada and I choose to take him around Ontario where I could, to allow him to have a good expierence in Canada.  He arrived in October, so the weather was a bit cold, but the colours were gorgeous. We decided to take a trip to Niagara Falls, where we spent the weekend. I went to school in Niagara Falls, so I have been there many times before, not expecting much from the trip. \\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}},{\\"key\\":\\"728k6\\",\\"text\\":\\"\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}},{\\"key\\":\\"79oud\\",\\"text\\":\\"\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}},{\\"key\\":\\"3k8rc\\",\\"text\\":\\" \\",\\"type\\":\\"atomic\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[{\\"offset\\":0,\\"length\\":1,\\"key\\":0}],\\"data\\":{}},{\\"key\\":\\"a3mbg\\",\\"text\\":\\"\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}},{\\"key\\":\\"7u5rt\\",\\"text\\":\\"\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}},{\\"key\\":\\"eu0ae\\",\\"text\\":\\"It turned out, I was in for a treat. Even though I had seen this place many times before, my travelling partner had really allowed me to see the beauty that I never took in before. As I watched him expierence everything for the first time, making comments on things he was observing, I realised I was observing these things through brand new eyes. The colours of Autum, the bright atmosphere of the attractions, the 'wow' factor of the falls and environment of the city itself. This trip made me come to the conclusion that even though you may have been to a place many times before, or even this may be a place that you call home, there will always be new things to discover. That you should always keep your eyes and mind open to observe the world around you and all that it has to offer. \\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}},{\\"key\\":\\"df8pe\\",\\"text\\":\\"\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}},{\\"key\\":\\"11lei\\",\\"text\\":\\"\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}},{\\"key\\":\\"6ddvh\\",\\"text\\":\\"\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}},{\\"key\\":\\"1jps5\\",\\"text\\":\\"\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}}],\\"entityMap\\":{\\"0\\":{\\"type\\":\\"IMAGE\\",\\"mutability\\":\\"IMMUTABLE\\",\\"data\\":{\\"src\\":\\"https://stravell-production.s3.amazonaws.com/pictures/ea5f10a1-10c3-4cfd-a607-982678db7159/49014533_290589528468380_8550961320021721088_n.jpg\\",\\"alignment\\":\\"center\\",\\"width\\":65}}}}",
          "createdAt": 1546199820574,
          "description": "Discovering the beauty of my country, my home.",
          "id": "-LV-kv5VGQDljYtMPt9v",
          "image": "https://stravell-production.s3.amazonaws.com/pictures/ea5f10a1-10c3-4cfd-a607-982678db7159/main",
          "s3FolderName": "ea5f10a1-10c3-4cfd-a607-982678db7159",
          "title": "Don't Forget Where You Come From.",
          "uid": "YYFEgQAdBuQIsox3ojncZtpSAmG3",
          "updatedAt": 1546221085951,
          "userName": "Jenna Jacquelyn",
          "userPhotoURL": "https://graph.facebook.com/10161147754845459/picture",
        },
      ]
    }
  />
</div>
`;

exports[`the component get passed an empty array of posts should render correctly and not display the filters component 1`] = `
<div>
  <PostList
    className="post-list post-list--no-border-top"
    posts={Array []}
  />
</div>
`;
